pkg/deploy/registry/deploylog/rest.go

```diff
@@ -0,0 +1,259 @@
+	deployments, err := r.Deploy.ListDeployments(ctx, deployutil.ConfigSelector(config.Name))
+	if err != nil {
+		return nil, err
+	}
+	if len(deployments.Items) == 0 {
+		return nil, errors.NewInternalError(fmt.Errorf("cannot find deployments for deploymentConfig %q", config.Name))
+	}
+	sort.Sort(deployutil.DeploymentsByLatestVersionDesc(deployments.Items))
+
+	// Find desired deployment
+	var target *kapi.ReplicationController
+	for _, deployment := range deployments.Items {
+		if desiredVersion == deployutil.DeploymentVersionFor(&deployment) {
+			target = &deployment
+			break
+		}
+	}
+	if target == nil {
+		desiredDeployment := deployutil.VersionedLabelForDeploymentConfig(config, desiredVersion)
+		glog.V(3).Infof("Cannot find desired deployment (%s) for %s", desiredDeployment, config.Name)
+		return nil, errors.NewNotFound("deployment", desiredDeployment)
+	}
+
+	// Check for deployment status; if it is new or pending, we will wait for it. If it is complete,
+	// the deployment completed successfully and the deployer pod will be deleted so we will return a
+	// success message. If it is running or failed, retrieve the log from the deployer pod.
+	status := deployutil.DeploymentStatusFor(target)
+	switch status {
+	case deployapi.DeploymentStatusNew, deployapi.DeploymentStatusPending:
+		if deployLogOpts.NoWait {
+			glog.V(4).Infof("Deployment %s is in %s state. No logs to retrieve yet.", deployutil.LabelForDeployment(target), status)
+			return &genericrest.LocationStreamer{}, nil
+		}
+		glog.V(4).Infof("Deployment %s is in %s state, waiting for it to start...", deployutil.LabelForDeployment(target), status)
+		lw := createDeploymentLW(ctx, r.Deploy, target)
+		if err := WaitForResource(&kapi.ReplicationController{}, lw, handleDeployment); err != nil {
+			return nil, err
+		}
+	case deployapi.DeploymentStatusComplete:
+		// Deployer pod has been deleted, no logs to retrieve
+		glog.V(4).Infof("Deployment %s was successful so the deployer pod is deleted. No logs to retrieve.", deployutil.LabelForDeployment(target))
+		return &genericrest.LocationStreamer{}, nil
+	}
+
+	deployPodName := deployutil.DeployerPodNameForDeployment(target.Name)
+	logOpts := &kapi.PodLogOptions{
+		Follow: deployLogOpts.Follow,
+	}
+	location, transport, err := pod.LogLocation(r.PodGetter, r.ConnectionInfo, ctx, deployPodName, logOpts)
+	if err != nil {
+		return nil, errors.NewBadRequest(err.Error())
+	}
+	return &genericrest.LocationStreamer{
+		Location:    location,
+		Transport:   transport,
+		ContentType: "text/plain",
+		Flush:       deployLogOpts.Follow,
+	}, nil
+}
+
+// REST implements GetterWithOptions
+var _ = rest.GetterWithOptions(&REST{})
+
+// createDeploymentLW creates a ListWatch scoped on deployments
+func createDeploymentLW(ctx kapi.Context, c *Client, deployment *kapi.ReplicationController) *cache.ListWatch {
+	return &cache.ListWatch{
+		ListFunc: func() (runtime.Object, error) {
+			return c.ListDeployments(ctx, labels.Everything())
```

Comment:
I have reworked `createDeploymentLW` to accept a label and a field selector instead of passing the deployment so this stays a `List`.